File: src/main/java/sharev/domain/connection/service/ConnectionService.java
package sharev.domain.connection.service;

import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import sharev.domain.card.entity.Card;
import sharev.domain.card.repository.CardRepository;
import sharev.domain.connection.entity.Connection;
import sharev.domain.connection.repository.ConnectionRepository;
import sharev.exception.CustomException;
import sharev.exception.ExceptionCode;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ConnectionService {

    private final ConnectionRepository connectionRepository;
    private final CardRepository cardRepository;

    @Transactional
    public void connect(UUID eventId, Long accountId, Long targetCardId) {
        Card card = cardRepository.findByGatheringIdAndAccountId(eventId, accountId)
                .orElseThrow(() -> new CustomException(ExceptionCode.CARD_NOT_FOUND));
        Card targetCard = cardRepository.findById(targetCardId)
                .orElseThrow(() -> new CustomException(ExceptionCode.CARD_NOT_FOUND));

        if (card.getId().equals(targetCard.getId())) {
            throw new CustomException(ExceptionCode.REGISTER_MYSELF);
        }

        List<Connection> connections = Connection.connect(card, targetCard);

        try {
            connectionRepository.saveAllAndFlush(connections);
        } catch (DataIntegrityViolationException e) {
            throw new CustomException(ExceptionCode.REGISTER_ALREADY);
        }
    }
}

File: src/main/kotlin/sharev/connection/application/service/ConnectionService.kt
package sharev.connection.application.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import sharev.card.application.port.outbound.LoadCardPort
import sharev.card.domain.exception.CardException
import sharev.card.domain.exception.CardExceptionCode as CardCode
import sharev.connection.domain.exception.ConnectionExceptionCode as ConnectionCode

@Service
@Transactional(readOnly = true)
class ConnectionService(
    private val loadCardPort: LoadCardPort,
    private val saveConnectionPort: sharev.connection.application.port.outbound.SaveConnectionPort,
) : sharev.connection.application.port.inbound.usecase.ConnectCardsUseCase {

    @Transactional
    override fun connect(command: sharev.connection.application.port.inbound.command.ConnectCardsCommand) {
        val card = loadCardPort.loadByGatheringAndAccount(command.gatheringId, command.accountId)
        val targetCard = loadCardPort.load(command.targetCardId)

        if (targetCard.gatheringId != command.gatheringId) {
            throw CardException(CardCode.CARD_NOT_FOUND)
        }

        if (card.id == targetCard.id) {
            throw _root_ide_package_.sharev.connection.domain.exception.ConnectionException(ConnectionCode.REGISTER_MYSELF)
        }

        saveConnectionPort.save(card.id, targetCard.id)
    }
}
